### PR TITLE
[COR-275] Optimizer Rule for ANY ==

### PR DESCRIPTION
**Pull request by cpjulia**
This PR implements the optimizer 'replace-any-eq-with-in' rule that maps queries with FILTER ... ANY == to queries with FILTER ... IN, , enabling index usage and other optimizations that don't happen with condition ANY ==. It happenes similarly to the rule 'replace-or-with-in'.

### Scope & Purpose

*(Please describe the changes in this PR for reviewers, motivation, rationale - **mandatory**)*

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made Obs.: NOT YET MADE MIGHT NEED ONE!
- [x] :books: documentation written (release notes, API changes, ...) https://github.com/arangodb/docs-hugo/pull/859
- [ ] Backports
  - [ ] Backport for 3.12.0: *(Please link PR)*
  - [ ] Backport for 3.11: *(Please link PR)*
  - [ ] Backport for 3.10: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [x] Docs PR: https://github.com/arangodb/docs-hugo/pull/859
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document: 

---

> [!NOTE]
> **Medium Risk**
> Touches AQL optimizer and AST comparison/deduplication logic, which can subtly change query plans and filter canonicalization; risk is mitigated by extensive new JS tests but may still affect edge-case semantics/perf (notably around subqueries and commutative comparisons).
> 
> **Overview**
> Adds a new optimizer rule, `replace-any-eq-with-in`, that rewrites `['a','b'] ANY == x.attr` into `x.attr IN ['a','b']` (only for deterministic arrays and attribute/reference-like operands), enabling index usage similarly to `replace-or-with-in`.
> 
> Improves condition optimization by introducing structural AST comparison (handling commutative operators, `IN` array ordering, references/attribute access/function calls) to remove duplicate conditions within AND-branches and to deduplicate identical OR-branches, while explicitly refusing to compare/deduplicate anything involving subqueries.
> 
> Extends constant folding/simplification behavior: `AstNode::isTrue`/`isFalse` now special-case empty `IN`/`NIN` arrays, single-value `IN` is optimized to `==` in more places, comparisons are normalized for `==`/`!=`, quantifiers can be stringified, and `removeUnnecessaryFiltersRule2` can replace constant-false filters with `NoResultsNode` after transformations. Tests are updated/added to cover duplicate detection edge cases and the new rule.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit a9a21476b99536626de3958df4822966f5482790. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by markuspf** [comment]
In some places it might be helpful for the readability of the code to use `TypedAstNodes` as introduced in https://github.com/arangodb/arangodb/pull/22001

File: arangod/Aql/Optimizer/Rule/OptimizerRuleReplaceAnyEqWithIn.cpp [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+    TRI_ASSERT(n->hasDependency());
+    auto const dep = n->getFirstDependency();
+
+    if (dep->getType() != EN::CALCULATION) {
+      continue;
+    }
+
+    auto fn = ExecutionNode::castTo<FilterNode const*>(n);
+    auto cn = ExecutionNode::castTo<CalculationNode*>(dep);
+    auto outVar = cn->outVariable();
```
Comment (maierlars) [inline arangod/Aql/Optimizer/Rule/OptimizerRuleReplaceAnyEqWithIn.cpp]: This might not be correct. The calculation for the FilterNode is not required to be the direct dependency. If you want to find the right calculation node for this FilterNode, consider using the function [`Plan::getVarSetBy`](https://github.com/arangodb/arangodb/blob/4185d910419a297c4d3d6720ce5a1d1cb901a3b7/arangod/Aql/ExecutionPlan.h#L209). :)

File: arangod/Aql/Optimizer/Rule/OptimizerRuleReplaceAnyEqWithIn.cpp [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+    AnySimplifier simplifier(*plan->getAst(), *plan.get());
+    auto newRoot = simplifier.simplify(root);
+
+    if (newRoot != root) {
+      auto expr = std::make_unique<Expression>(plan->getAst(), newRoot);
+
+      TRI_IF_FAILURE("OptimizerRules::replaceAnyEqWithInRuleOom") {
+        THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+      }
```
Comment (maierlars) [inline arangod/Aql/Optimizer/Rule/OptimizerRuleReplaceAnyEqWithIn.cpp]: there is no need to test out of memory here.

**Review by maierlars — CHANGES_REQUESTED** [review]
This PR wants to be three different PRs. Please separate them, so that the review process becomes bearable.
